### DISCUSSION TITLE
"Failed to load url" error if package.json and vite.config.ts is not in parent directory to the tested code

### DISCUSSION THREAD

**Discussion by trumko** [vitest-dev/vitest#5279] [Q&A]
### Describe the bug

I am trying to migrate to the vitest+React Testing Library and have issue which could be related to the file structure. In case if package.json and vite.config.ts file is not in the parent directory to the tested code I got the error:

```
 FAIL  ../src/App.test.tsx [ ../src/App.test.tsx ]
Error: Failed to load url D:/vit/react-testing-lib/src/App.test.tsx (resolved id: D:/vit/react-testing-lib/src/App.test.tsx). Does the file exist?
 ❯ loadAndTransform node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:53601:21
```

### Reproduction

For a reproduction purpose the [react-testing-lib example](https://github.com/vitest-dev/vitest/tree/main/examples/react-testing-lib) can be used.

react-testing-lib example "out of box" works great and tests are run correctly.

However if I change the file structure from this:
```
react-testing-lib
├──  index.html
├──  package.json
├──  tconfig.json
├──  vite.config.ts
└──  src
  └── [React components and tests]
```

to this (I need similar structure on my real project):
```
react-testing-lib
├── build
  ├──  package.json
  ├──  tconfig.json
  ├──  vite.config.ts
  └── setup.ts
└── src
  └── [React components and tests]
```
and update vite.config.ts to properly handle new structure:
```
/// <reference types="vitest" />
/// <reference types="vite/client" />

import react from '@vitejs/plugin-react'
import { defineConfig } from 'vite'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  test: {
    include: ['../src/**/*.{test,spec}.?(c|m)[jt]s?(x)'], // *
    globals: true,
    environment: 'jsdom',
    setupFiles: './setup.ts', // **
    css: true,
  },
})

/*
    * added this include to handle new file structure
    ** was forced to change it for now since also get the error "Error: Failed to load url D:/vit/react-testing-lib/src/test/setup.ts (resolved id: D:/vit/react-testing-lib/src/test/setup.ts). Does the file exist?"
*/
```
I got the errors like:
```
 npm test

> test
> vitest


 DEV  v1.3.1 D:/vit/react-testing-lib/build

 ❯ ../src/components/Spinner.test.tsx (0)
 ❯ ../src/App.test.tsx (0)
 ❯ ../src/components/input.test.tsx (0)
 ❯ ../src/hooks/useCounter.test.ts (0)

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Suites 4 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  ../src/App.test.tsx [ ../src/App.test.tsx ]
Error: Failed to load url D:/vit/react-testing-lib/src/App.test.tsx (resolved id: D:/vit/react-testing-lib/src/App.test.tsx). Does the file exist?
 ❯ loadAndTransform node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:53601:21

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/4]⎯

 FAIL  ../src/components/input.test.tsx [ ../src/components/input.test.tsx ]
Error: Failed to load url D:/vit/react-testing-lib/src/components/input.test.tsx (resolved id: D:/vit/react-testing-lib/src/components/input.test.tsx). Does the file exist?
 ❯ loadAndTransform node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:53601:21

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/4]⎯

 FAIL  ../src/components/Spinner.test.tsx [ ../src/components/Spinner.test.tsx ]
Error: Failed to load url D:/vit/react-testing-lib/src/components/Spinner.test.tsx (resolved id: D:/vit/react-testing-lib/src/components/Spinner.test.tsx). Does the file exist?
 ❯ loadAndTransform node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:53601:21

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[3/4]⎯

 FAIL  ../src/hooks/useCounter.test.ts [ ../src/hooks/useCounter.test.ts ]
Error: Failed to load url D:/vit/react-testing-lib/src/hooks/useCounter.test.ts (resolved id: D:/vit/react-testing-lib/src/hooks/useCounter.test.ts). Does the file exist?
 ❯ loadAndTransform node_modules/vite/dist/node/chunks/dep-jDlpJiMN.js:53601:21

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[4/4]⎯

 Test Files  4 failed (4)
      Tests  no tests
   Start at  20:02:36
   Duration  889ms (transform 30ms, setup 758ms, collect 0ms, tests 0ms, environment 1.60s, prepare 422ms)
```

Would be grateful to hear your thoughts. Thank you.

### System Info

```shell
System:
    OS: Windows 11 10.0.22631
  Binaries:
    Node: 20.11.1 - C:\Program Files\nodejs\node.EXE
    npm: 10.2.4 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Chromium (121.0.2277.128)
    Internet Explorer: 11.0.22621.1
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by hi-ogawa** [comment]
Your setup looks very unusual and especially if you have `(root)/build/package.json` outside of `(root)/src`, then I don't think `(root)/src` can access `(root)/build/node_modules` in any tooling (not only Vitest, but I suppose you cannot run Vite app in such structure).

**Comment by trumko** [comment]
@hi-ogawa I'd say this is the structure my project historically has. We don't use Vite. And Jest tool works pretty well with this structure. Now I am investigating possibility to migrate to the Vitest and wondering if it possible to use the Vitest without changing project structure.

**Reply by sheremet-va** [reply]
This is cause by the security check. Try disabling [`server.fs.strict`](https://vitejs.dev/config/server-options.html#server-fs-strict) or adding the outer folder to `server.fs.allow`

**Reply by trumko** [reply]
Thank you, disabling [server.fs.strict](https://vitejs.dev/config/server-options.html#server-fs-strict)  solved the issue. But looks like now vitest has dificluties with resolving some package paths in scr folder:
`Error: Failed to resolve import "@testing-library/jest-dom" from "../src/test/setup.ts". Does the file exist?`

**Reply by hi-ogawa** [reply]
@trumko I think that's somewhat expected by the reason I wrote this above:
>  I don't think `(root)/src` can access `(root)/build/node_modules`

To understand your situation better, I think it would be great if you can provide a example project where jest is working with such project structure. Then, maybe we can find a similar way to make it work on Vitest too.

**Reply by sheremet-va** [reply]
Maybe changing the [root](https://vitest.dev/config/#root) might help.

**Reply by trumko** [reply]
So I played around with `root` (and correspondently changing `test.include` and `test.setupFiles` paths) but still facing the same issue

**Reply by hi-ogawa** [reply]
That's the limit of what we can guess, so we'll need a runnable/failing reproduction to suggest more precise idea.
Can you provide a reproduction on stackblitz?

**Reply by trumko** [reply]
@hi-ogawa sure - [reproduction on stackblitz](https://stackblitz.com/edit/vitejs-vite-gmmge5?file=build%2Fvite.config.ts)
Steps:
```
cd ./build
npm install
npm test
```

![2024-03-06_10h48_13](https://github.com/vitest-dev/vitest/assets/12519281/73579a9c-d552-42dc-a915-8c83dbf3f61d)

**Reply by hi-ogawa** [reply]
Thanks for the reproduction. Hmm, I might be missing something, but this looks super wild to me.
Quick questions:
- It looks like `vite` doesn't run either in this case. What tool do you use to run this app locally (other than jest)?
- Can you maybe create a example where Jest runs with this setup?
